File: ecommerce/backend/app/schemas/promotion.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict

from app.schemas.common import TimestampMixin


class PromotionBase(BaseModel):
    name: str
    description: str | None = None
    promotion_type: str
    discount_type: str | None = None
    discount_value: float | None = None
    product_id: int | None = None
    category: str | None = None
    min_quantity: int | None = None
    free_product_id: int | None = None
    min_order_amount: float | None = None
    max_discount: float | None = None
    loyalty_multiplier: float | None = None
    start_date: datetime
    end_date: datetime
    is_active: bool = True


class PromotionCreate(PromotionBase):
    pass


class PromotionUpdate(BaseModel):
    name: str | None = None
    description: str | None = None
    promotion_type: str | None = None
    discount_type: str | None = None
    discount_value: float | None = None
    product_id: int | None = None
    category: str | None = None
    min_quantity: int | None = None
    free_product_id: int | None = None
    min_order_amount: float | None = None
    max_discount: float | None = None
    loyalty_multiplier: float | None = None
    start_date: datetime | None = None
    end_date: datetime | None = None
    is_active: bool | None = None


class PromotionResponse(PromotionBase, TimestampMixin):
    id: int

    model_config = ConfigDict(from_attributes=True)


class PromotionValidateRequest(BaseModel):
    promotion_id: int
    order_subtotal: float
    product_ids: list[int] | None = None


class PromotionValidateResponse(BaseModel):
    is_applicable: bool
    discount_amount: float = 0
    discount_type: str | None = None
    description: str | None = None

File: mobile/shared_core/lib/src/models/promotion.dart
class Promotion {
  final int id;
  final String name;
  final String? description;
  final String promotionType;
  final String? discountType;
  final double? discountValue;
  final int? productId;
  final String? category;
  final double? minOrderAmount;
  final double? maxDiscount;
  final bool isActive;
  final String startDate;
  final String endDate;

  Promotion({
    required this.id,
    required this.name,
    this.description,
    required this.promotionType,
    this.discountType,
    this.discountValue,
    this.productId,
    this.category,
    this.minOrderAmount,
    this.maxDiscount,
    this.isActive = true,
    required this.startDate,
    required this.endDate,
  });

  factory Promotion.fromJson(Map<String, dynamic> json) {
    return Promotion(
      id: json['id'],
      name: json['name'] ?? '',
      description: json['description'],
      promotionType: json['promotion_type'] ?? '',
      discountType: json['discount_type'],
      discountValue: (json['discount_value'] as num?)?.toDouble(),
      productId: json['product_id'],
      category: json['category'],
      minOrderAmount: (json['min_order_amount'] as num?)?.toDouble(),
      maxDiscount: (json['max_discount'] as num?)?.toDouble(),
      isActive: json['is_active'] ?? true,
      startDate: json['start_date'] ?? '',
      endDate: json['end_date'] ?? '',
    );
  }

  String get discountLabel {
    if (discountType == 'percentage' && discountValue != null) {
      return '${discountValue!.toInt()}% OFF';
    }
    if (discountValue != null) {
      return 'KSh ${discountValue!.toInt()} OFF';
    }
    return promotionType;
  }
}
